Use logrus for structured logging

ankushagarwal opened the pull request:
* Move from glog to sirupsen/logrus for logging
* Add a new flag json_log_format
* Refactor all log statements to use logrus
* Use Info level for log level V(1) and below
* Use Debug level for log level V(2) and above
* Tested locally

Addresses https://github.com/kubeflow/tf-operator/issues/24

Sample logs
```
{"filename":"app/server.go:54","level":"info","msg":"EnvKubeflowNamespace not set, use default namespace","time":"2018-02-27T18:25:18-08:00"}
{"filename":"app/server.go:59","level":"info","msg":"[Version: 0.3.0+git Git SHA: Not provided. Go Version: go1.9.3 Go OS/Arch: darwin/amd64]","time":"2018-02-27T18:25:18-08:00"}
{"filename":"app/server.go:145","level":"info","msg":"No controller_config_file provided; using empty config.","time":"2018-02-27T18:25:18-08:00"}
{"filename":"controller/controller.go:110","level":"info","msg":"Setting up event handlers","time":"2018-02-27T18:25:18-08:00"}
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/tf-operator/416)
<!-- Reviewable:end -->

jlewi commented:
nit: Could you include some sample log output in the PR description?

jlewi commented:
Does the log output include line number or just filename?

ankushagarwal commented:
Added sample log output.

The log output includes line numbers as well.

gaocegege commented:
Thanks for your contribution :tada: 

There is two concerns: 

- Will vendor's log printed if we use logrus but most of components in k8s uses glog? For example, https://github.com/kubeflow/tf-operator/blob/master/vendor/k8s.io/apimachinery/pkg/api/meta/meta.go#L596.
- Should we update the code in client/ manually to keep it use logrus instead of glog? The generated code uses glog by default.

gaocegege commented:
/retest

jlewi commented:
/assign jlewi
/assign @gaocegege

ankushagarwal commented:
Created #424

jlewi commented:
I'd suggest resyncing to see if that fixes the tests.
/lgtm
/hold
   Lets wait for @gaocegege  to review

ankushagarwal commented:
/retest

gaocegege commented:
I will have a try and submit my review today.

gaocegege commented:
I have observation to share with you @ankushagarwal @jlewi , I am not sure how to deal with it.

Now we use flag package to support command line flags, and glog also uses it by default. Then you can see our binary have more flags than we thing although we use logrus instead of glog:

```
➜  tf-operator git:(416) ✗ ./tf-operator -h               
Usage of ./tf-operator:
  -alsologtostderr
    	log to standard error as well as files
  -chaos-level int
    	DO NOT USE IN PRODUCTION - level of chaos injected into the TFJob created by the operator. (default -1)
  -controller-config-file string
    	Path to file containing the controller config.
  -gc-interval duration
    	GC interval (default 10m0s)
  -json-log-format
    	Set true to use json style log format. Set false to use plaintext style log format (default true)
  -log_backtrace_at value
    	when logging hits line file:N, emit a stack trace
  -log_dir string
    	If non-empty, write log files in this directory
  -logtostderr
    	log to standard error instead of files
  -stderrthreshold value
    	logs at or above this threshold go to stderr
  -v value
    	log level for V logs
  -version
    	Show version and quit
  -vmodule value
    	comma-separated list of pattern=N settings for file-filtered logging
```

There are some pros and cons:

- We can support vendor and client's glog, since we have glog's flags
- But the users may be confused since the tf-operator outputs logs regardless of the flag `-logtostderr`

ankushagarwal commented:
@gaocegege I have changed `pkg/client/clientset/versioned/clientset.go` which does not seem to have auto generated. Also looks like `./hack/update-codegen.sh` is [broken](https://pastebin.com/raw/qaGDbPsD)

I have fixed the `-version` logging issue.

gaocegege commented:
/approve

@ankushagarwal 

pkg/client is all auto-generated by kubernetes/code-generator.

update-codegen.sh works for me, it seems that the script uses the wrong code-generator in kubernetes repo. 

But no worry about it, I can open a PR to fix the clientset.

Thanks for your awesome contribution and patience! The log problem has been bothering us for long time.

gaocegege commented:
/retest

gaocegege commented:
@ankushagarwal 

Sorry, the CI failed because we use `"github.com/sirupsen/logrus"` but the vendor's logrus is `"github.com/onrik/logrus"`

Could you fix it?

ankushagarwal commented:
@gaocegege : Can you link me to the failure?

We use both `sirupsen/logrus` and `onrik/logrus`

gaocegege commented:
@ankushagarwal 

Sure

http://testing-argo.kubeflow.io/timeline/kubeflow-test-infra/kubeflow-tf-operator-presubmit-tfjob-e2e-416-c2e0055-119-b903;tab=workflow;node=kubeflow-tf-operator-presubmit-tfjob-e2e-416-c2e0055-119-b903-4276748402:logs

ankushagarwal commented:
Added sirupsen/logrus as a glide dependency - hopefully build should pass now

ankushagarwal commented:
/retest
